### PR TITLE
feat: type definitions and documentation for separated type provider.

### PR DESCRIPTION
**Pull request by Bram-dc**
Fix for discussion: https://github.com/fastify/fastify/discussions/5311
An update on a previous PR

This improved PR includes:
- Breaking changes with separate providers for both the validation and serialization stages
- Docs for the new TypeProvider interface
- A test

### CODE REVIEW

**Comment by Bram-dc** [comment]
Will this be included in the new 5.0.0 release? I could start to create PRs for all the type-providers to support this changes.

**Comment by mcollina** [comment]
cc @fastify/typescript what do you think? Should we do this?

**Comment by fox1t** [comment]
I encountered this same limitation recently while working with the Zod-type provider. I am +1 for this landing. Since I haven't worked recently on the type provider types, though, I am not sure how much breaking this PR is.

**Review by fox1t — APPROVED** [review]

**Comment by mcollina** [comment]
@Bram-dc could you start preparing the matching PRs for our type providers? So we can land this for v5.

**Comment by Bram-dc** [comment]
@mcollina Could we have a npm publish to the "next" tag so that we can try this change in the plugins pipelines?

**Comment by mcollina** [comment]
This conflicts now, can you rebase?

**Comment by Bram-dc** [comment]
I think the conflict is resolved now, I did not completely understand the conflict.

**Comment by Bram-dc** [comment]
I think I messed something up, the files changed count is up.

File: test/types/type-provider.test-d.ts [hunk 1]
```diff
@@ -1053,7 +1062,7 @@ expectAssignable(server.withTypeProvider<InlineHandlerProvider>().get(
 // Handlers: Auxiliary
 // -------------------------------------------------------------------
 
-interface AuxiliaryHandlerProvider extends FastifyTypeProvider { output: this['input'] }
```
Comment (Bram-dc) [inline test/types/type-provider.test-d.ts:1056]: Why was this changed from `{ output: 'auxiliary' }`

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by Bram-dc** [comment]
Could we have a "5.0.0-alpha.2" like [5.0.0-aplha.1](https://www.npmjs.com/package/fastify/v/5.0.0-aplha.1)
